examples/src/main/java/com/squareup/moshi/recipes/Polymorphic.java

```diff
@@ -0,0 +1,126 @@
+        .add(new CardAdapter())
+        .build();
+    String handJson = "{\n"
+        + "    \"kind\": \"blackjack-hand\",\n"
+        + "    \"data\": {\n"
+        + "      \"hidden_card\": \"6S\",\n"
+        + "      \"visible_cards\": [\n"
+        + "        \"4C\",\n"
+        + "        \"AH\"\n"
+        + "      ]\n"
+        + "    }\n"
+        + "  }";
+    String playerJson = "{\n"
+        + "    \"kind\": \"player\",\n"
+        + "    \"data\": {\n"
+        + "      \"username\": \"jesse\",\n"
+        + "      \"lucky number\": 32\n"
+        + "    }\n"
+        + "  }";
+    JsonAdapter<BlackjackHand> handAdapter = moshi.adapter(BlackjackHand.class);
+    JsonAdapter<Player> playerAdapter = moshi.adapter(Player.class);
+    System.out.println(handAdapter.fromJson(handJson));
+    System.out.println(playerAdapter.fromJson(playerJson));
+    System.out.println(handAdapter.toJson(
+        new BlackjackHand(new Card('A', Suit.SPADES), Collections.<Card>emptyList())));
+    System.out.println(playerAdapter.toJson(new Player("jake", -1)));
+  }
+
+  public static final class PolymorphicBlackjackJsonAdapter extends JsonAdapter<Object> {
+    public static final JsonAdapter.Factory FACTORY = new Factory() {
+      @Override
+      public JsonAdapter<?> create(Type type, Set<? extends Annotation> annotations, Moshi moshi) {
+        if (type != BlackjackHand.class && type != Player.class) {
+          return null;
+        }
+        JsonAdapter<BlackjackHand> handAdapter =
+            moshi.nextAdapter(this, BlackjackHand.class, Collections.<Annotation>emptySet());
+        JsonAdapter<Player> playerAdapter =
+            moshi.nextAdapter(this, Player.class, Collections.<Annotation>emptySet());
+        return new PolymorphicBlackjackJsonAdapter(handAdapter, playerAdapter);
+      }
+    };
+
+    private final JsonAdapter<BlackjackHand> handAdapter;
+    private final JsonAdapter<Player> playerAdapter;
+
+    PolymorphicBlackjackJsonAdapter(JsonAdapter<BlackjackHand> handAdapter,
+        JsonAdapter<Player> playerAdapter) {
+      this.handAdapter = handAdapter;
+      this.playerAdapter = playerAdapter;
+    }
+
+    @Override public Object fromJson(JsonReader reader) throws IOException {
+      // noinspection unchecked
+      Map<String, Object> value = (Map<String, Object>) reader.readJsonValue();
+      String type = (String) value.get("kind");
+      Object data = value.get("data");
+      switch (type) {
+        case "blackjack-hand":
+          return handAdapter.fromJsonValue(data);
+        case "player":
+          return playerAdapter.fromJsonValue(data);
+        default:
+          throw new JsonDataException("Unexpected type: " + type);
+      }
+    }
+
+    @Override public void toJson(JsonWriter writer, Object value) throws IOException {
+      Class<?> cls = value.getClass();
+      writer.beginObject();
+      writer.name("kind");
+      String type;
+      JsonAdapter adapter;
```

Comment:
Well, if you do that, the `adapter.toJson` call doesn't compile. If you're trying to avoid using raw types, you can get around it with something similar to:
```java
@Override public void toJson(JsonWriter writer, Object value) throws IOException {
  Class<?> cls = value.getClass();
  writer.beginObject();
  writer.name("kind");
  if (cls == BlackjackHand.class) {
      writeTypeAndValueJson(writer, "blackjack-hand", handAdapter, value);
  } else if (cls == Player.class) {
      writeTypeAndValueJson(writer, "player", playerAdapter, value);
  } else {
    throw new AssertionError(cls + " is not BlackJackHand.class or Player.class.");
  }
  writer.endObject();
}

private static <T> void writeTypeAndValueJson(JsonWriter writer, String type, JsonAdapter<T> adapter, Object value) throws IOException {
    writer.value(type);
    writer.name("data");
    //noinspection unchecked
    adapter.toJson(writer, (T) value);
}
```